mylyn.tasks/connectors/gitlab/org.eclipse.mylyn.gitlab.core/src/org/eclipse/mylyn/internal/gitlab/core/GitlabRepositoryConnector.java

```diff
@@ -0,0 +1,550 @@
+import org.eclipse.mylyn.tasks.core.data.TaskMapper;
+import org.eclipse.mylyn.tasks.core.sync.ISynchronizationSession;
+
+import com.google.common.base.Objects;
+import com.google.common.base.Optional;
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.CacheLoader;
+import com.google.common.cache.LoadingCache;
+import com.google.common.collect.Lists;
+import com.google.common.util.concurrent.ExecutionError;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.common.util.concurrent.UncheckedExecutionException;
+
+public class GitlabRepositoryConnector extends AbstractRepositoryConnector {
+
+    public class RepositoryKey {
+	private final TaskRepository repository;
+
+	public RepositoryKey(@NonNull TaskRepository repository) {
+	    super();
+	    this.repository = repository;
+	}
+
+	public TaskRepository getRepository() {
+	    return repository;
+	}
+
+	@Override
+	public int hashCode() {
+	    return repository.hashCode();
+	}
+
+	@Override
+	public boolean equals(Object obj) {
+	    if (this == obj) {
+		return true;
+	    }
+	    if (obj == null || getClass() != obj.getClass()) {
+		return false;
+	    }
+	    return this.repository.equals(((RepositoryKey) obj).getRepository());
+	}
+    }
+
+    private static final ThreadLocal<IOperationMonitor> context = new ThreadLocal<IOperationMonitor>();
+
+    private final LoadingCache<RepositoryKey, Optional<GitlabConfiguration>> configurationCache;
+    private SimpleDateFormat simpleFormatter = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'");
+    public static Duration CONFIGURATION_CACHE_EXPIRE_DURATION = new Duration(7, TimeUnit.DAYS);
+    public static Duration CONFIGURATION_CACHE_REFRESH_AFTER_WRITE_DURATION = new Duration(1, TimeUnit.DAYS);
+    public static Duration CLIENT_CACHE_DURATION = new Duration(24, TimeUnit.HOURS);
+
+    private boolean ignoredProperty(String propertyName) {
+	if (propertyName.equals(RepositoryLocation.PROPERTY_LABEL) || propertyName.equals(TaskRepository.OFFLINE)
+		|| propertyName.equals(IRepositoryConstants.PROPERTY_ENCODING)
+		|| propertyName.equals(TaskRepository.PROXY_HOSTNAME) || propertyName.equals(TaskRepository.PROXY_PORT)
+		|| propertyName.equals("org.eclipse.mylyn.tasklist.repositories.savePassword") //$NON-NLS-1$
+		|| propertyName.equals("org.eclipse.mylyn.tasklist.repositories.proxy.usedefault") //$NON-NLS-1$
+		|| propertyName.equals("org.eclipse.mylyn.tasklist.repositories.proxy.savePassword") //$NON-NLS-1$
+		|| propertyName.equals("org.eclipse.mylyn.tasklist.repositories.proxy.username") //$NON-NLS-1$
+		|| propertyName.equals("org.eclipse.mylyn.tasklist.repositories.proxy.password") //$NON-NLS-1$
+		|| propertyName.equals("org.eclipse.mylyn.tasklist.repositories.proxy.enabled")) { //$NON-NLS-1$
+	    return true;
+	}
+	return false;
+    }
+
+    abstract class ListenableFutureJob<V> extends Job implements ListenableFuture<V> {
```

Comment:
We use com.google.common.cache.LoadingCache but do not need to overwrite CacheLoader().reload() so we can delete ListenableFutureJob. This will be in the next commit.